client/src/main/java/io/pravega/client/stream/Transaction.java

```diff
@@ -70,9 +70,11 @@
      * Send a transaction heartbeat and increase transaction's timeout by lease amount of milliseconds.
      *
      * @param lease Additional amount of time in milliseconds by which to increase transaction's timeout.
+     * @param switchOver Boolean indicating whether the previous successful ping request was performed on a different
+     *                   controller process by the client.
      * @throws PingFailedException Ping failed.
      */
-    void ping(long lease) throws PingFailedException;
+    void ping(long lease, boolean switchOver) throws PingFailedException;
```

Comment:
@bhargav-gulavani i think since O(active txns) >> O(active streams) so using stable storage (in our case zk) for that many writes and reads for every ping may not be ideal. 

Correct me if i am wrong but following is the objective:
1. We want to abort transactions when timeouts elapse. 
2. We want an ability to keep renewing the lease so that the timer is reset everytime we receive pings. 
3. Since pings are received on WAN, we should have sufficient margins keeping large latencies in mind. 

So if the current solution uses stable store to fence and take exclusive rights upon receiving 'first ping' and if client can send pings in a round robin manner to controller instances, then we will end up updating stable storage O(num of pings) * O(num of txns). 

In such case we might as well use a stable storage to store pings. 
I would argue that pravega stream is better suited for such high throughput than zk. 
Also, with pravega streams readergroup we will get failover handling inbuilt with our reader failure detection.